File: hall_of_fame/wi-seong-cheol/BOJ/0x09 BFS/1600.swift
//
//  1600.swift
//  wi-seong-cheol
//
//  Created by wi_seong on 2023/03/24.
//

/*
 [Input]
 1
 4 4
 0 0 0 0
 1 0 0 0
 0 0 1 0
 0 1 0 0
 [Output]
 4
 */
// 시간복잡도: O(n * m)
// 최악시간: 20,000
// 난이도: Gold 3
// Timer: 1h
// Url: https://www.acmicpc.net/problem/1600

import Foundation

// 변수 선언
var k = 0, w = 0, h = 0
var board = [[Int]]()
var dx = [-1, 1, 0, 0]
var dy = [0, 0, -1, 1]
var horseDx = [-2, -1, -2, -1, 1, 2, 2, 1]
var horseDy = [-1, -2, 1, 2, -2, -1, 1, 2]

// 입력
k = Int(String(readLine()!))!
let wh = readLine()!.split(separator: " ").map{Int(String($0))!}
(w, h) = (wh[0], wh[1])
for _ in 0..<h{
    board.append(readLine()!.split(separator: " ").map{Int(String($0))!})
}

// 풀이
func OOB(_ x: Int, _ y: Int) -> Bool {
    return x < 0 || x >= h || y < 0 || y >= w
}

func bfs() -> Int{
    var queue = [(Int, Int, Int, Int)]()//x, y, 이동횟수, k
    var visited = Array(repeating: Array(repeating: Array(repeating: false, count: 31), count: 201), count: 201)
    queue.append((0, 0, 0, k))
    visited[0][0][k] = true
    var result = -1
    var idx = 0
    while queue.count > idx{
        let (x, y, move, k) = queue[idx]
        idx += 1
        if x == h - 1 && y == w - 1 {
            result = move
            break
        }

        if k > 0 {
            for i in 0..<8{
                let nx = horseDx[i] + x
                let ny = horseDy[i] + y

                if OOB(nx, ny) { continue }
                if board[nx][ny] == 1 || visited[nx][ny][k - 1] { continue }
                queue.append((nx, ny, move + 1, k - 1))
                visited[nx][ny][k - 1] = true
            }
        }
        for i in 0..<4{
            let nx = x + dx[i]
            let ny = y + dy[i]
            if OOB(nx, ny) { continue }
            if board[nx][ny] == 1 || visited[nx][ny][k] { continue }
            queue.append((nx, ny, move + 1, k))
            visited[nx][ny][k] = true
        }


    }
    return result
}

// 출력
print(bfs())

File: hall_of_fame/wi-seong-cheol/BOJ/0x09 BFS/1600.cpp
//
//  1600.cpp
//  main
//
//  Created by wi_seong on 2022/11/14.
//

/*
 [Input]
 1
 4 4
 0 0 0 0
 1 0 0 0
 0 0 1 0
 0 1 0 0
 [Output]
 4
 */
// 시간복잡도: O(n^2)
// 최악시간: 1,200,000
// 난이도: Gold 3
// Timer: 20m
// Url: https://www.acmicpc.net/problem/1600

#include <iostream>
#include <queue>
#include <algorithm>

using namespace std;
int k, w, h;
int dist[201][201][31];
int board[201][201];
int dx[] = {0, 1, 0, -1, 1, 2, 2, 1, -1, -2, -2, -1};
int dy[] = {1, 0, -1, 0, 2, 1, -1, -2, -2, -1, 1, 2};

bool OOB(int x, int y) { return x < 0 || x > h || y < 0 || y > w; }

void bfs() {
    queue<tuple<int, int, int>> Q;
    Q.push({0, 0, 0});
    dist[0][0][0] = 0;
    
    while(!Q.empty()) {
        auto [x, y, z] = Q.front(); Q.pop();
        if(x == h - 1 && y == w - 1) {
            cout << dist[x][y][z] << '\n';
            return;
        }
        for(int dir = 0; dir < 12; dir++) {
            int nx = x + dx[dir];
            int ny = y + dy[dir];
            int nz = z;
            if(OOB(nx, ny) || board[nx][ny] == 1) continue;
            if(dir >= 4) nz++;
            if(dist[nx][ny][nz] != -1 || nz > k) continue;
            dist[nx][ny][nz] = dist[x][y][z] + 1;
            Q.push({nx, ny, nz});
        }
    }
    cout << -1;
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    cin >> k >> w >> h;
    for(int i = 0; i < h; i++) {
        for(int j = 0; j < w; j++) {
            fill(dist[i][j], dist[i][j] + 31, -1);
            cin >> board[i][j];
        }
    }
    bfs();
    
    return 0;
}
